File: agents/python/transport/tcp.py
import socket
import struct
import time
import logging

log = logging.getLogger("phantom.transport.tcp")


class TcpTransport:
    """TCP transport with 4-byte length-prefix framing."""

    def __init__(self, host: str, port: int, timeout: float = 30.0,
                 retry_count: int = -1, retry_delay: float = 10.0):
        self.host = host
        self.port = port
        self.timeout = timeout
        self.retry_count = retry_count
        self.retry_delay = retry_delay
        self._sock: socket.socket | None = None
        self._connected = False

    @property
    def connected(self) -> bool:
        return self._connected

    def connect(self) -> bool:
        attempts = 0
        while self.retry_count == -1 or attempts < self.retry_count:
            try:
                self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self._sock.settimeout(self.timeout)
                self._sock.connect((self.host, self.port))
                self._connected = True
                log.info(f"Connected to {self.host}:{self.port}")
                return True
            except OSError as e:
                attempts += 1
                log.warning(f"Connection failed (attempt {attempts}): {e}")
                self._cleanup_socket()
                if self.retry_count != -1 and attempts >= self.retry_count:
                    return False
                time.sleep(self.retry_delay)
        return False

    def send(self, data: bytes) -> bool:
        if not self._connected or not self._sock:
            return False
        try:
            frame = struct.pack(">I", len(data)) + data
            self._sock.sendall(frame)
            return True
        except OSError as e:
            log.error(f"Send failed: {e}")
            self._connected = False
            return False

    def recv(self) -> bytes | None:
        if not self._connected or not self._sock:
            return None
        try:
            length_data = self._recv_exact(4)
            if not length_data:
                return None
            length = struct.unpack(">I", length_data)[0]
            if length > 10 * 1024 * 1024:
                log.error(f"Message too large: {length} bytes")
                return None
            return self._recv_exact(length)
        except OSError as e:
            log.error(f"Recv failed: {e}")
            self._connected = False
            return None

    def _recv_exact(self, n: int) -> bytes | None:
        buf = bytearray()
        while len(buf) < n:
            chunk = self._sock.recv(n - len(buf))
            if not chunk:
                self._connected = False
                return None
            buf.extend(chunk)
        return bytes(buf)

    def disconnect(self):
        self._cleanup_socket()
        log.info("Disconnected")

    def _cleanup_socket(self):
        self._connected = False
        if self._sock:
            try:
                self._sock.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

File: agents/go/transport/tcp.go
package transport

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"sync"
	"time"
)

const (
	maxMessageSize  = 10 * 1024 * 1024 // 10MB
	connectTimeout  = 10 * time.Second
	defaultRetries  = 3
	defaultRetryDel = 5 * time.Second
)

// TCPTransport implements length-prefixed TCP communication.
type TCPTransport struct {
	host       string
	port       int
	conn       net.Conn
	mu         sync.Mutex
	retries    int
	retryDelay time.Duration
	connected  bool
}

func NewTCPTransport(host string, port int) *TCPTransport {
	return &TCPTransport{
		host:       host,
		port:       port,
		retries:    defaultRetries,
		retryDelay: defaultRetryDel,
	}
}

func (t *TCPTransport) Connect() error {
	t.mu.Lock()
	defer t.mu.Unlock()

	addr := fmt.Sprintf("%s:%d", t.host, t.port)
	dialer := net.Dialer{Timeout: connectTimeout}

	var lastErr error
	for i := 0; i <= t.retries; i++ {
		conn, err := dialer.Dial("tcp", addr)
		if err == nil {
			t.conn = conn
			t.connected = true
			return nil
		}
		lastErr = err
		if i < t.retries {
			time.Sleep(t.retryDelay)
		}
	}
	return fmt.Errorf("tcp connect failed after %d retries: %w", t.retries, lastErr)
}

// Send writes a 4-byte big-endian length prefix followed by the payload.
func (t *TCPTransport) Send(data []byte) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if !t.connected || t.conn == nil {
		return fmt.Errorf("not connected")
	}

	if len(data) > maxMessageSize {
		return fmt.Errorf("message too large: %d > %d", len(data), maxMessageSize)
	}

	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(data)))

	if _, err := t.conn.Write(header); err != nil {
		t.connected = false
		return fmt.Errorf("send header: %w", err)
	}
	if _, err := t.conn.Write(data); err != nil {
		t.connected = false
		return fmt.Errorf("send payload: %w", err)
	}
	return nil
}

// Recv reads a 4-byte big-endian length prefix then reads that many bytes.
func (t *TCPTransport) Recv() ([]byte, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	if !t.connected || t.conn == nil {
		return nil, fmt.Errorf("not connected")
	}

	header := make([]byte, 4)
	if _, err := io.ReadFull(t.conn, header); err != nil {
		t.connected = false
		return nil, fmt.Errorf("recv header: %w", err)
	}

	length := binary.BigEndian.Uint32(header)
	if length > maxMessageSize {
		t.connected = false
		return nil, fmt.Errorf("message too large: %d", length)
	}

	buf := make([]byte, length)
	if _, err := io.ReadFull(t.conn, buf); err != nil {
		t.connected = false
		return nil, fmt.Errorf("recv payload: %w", err)
	}
	return buf, nil
}

func (t *TCPTransport) Disconnect() {
	t.mu.Lock()
	defer t.mu.Unlock()

	if t.conn != nil {
		t.conn.Close()
		t.conn = nil
	}
	t.connected = false
}

func (t *TCPTransport) Connected() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.connected
}
